mqtt/mqtt-bridge/src/pump.rs

```diff
@@ -0,0 +1,283 @@
+/// After connection there are two simultaneous processes:
+/// 1) persist incoming messages into an ingress store to be used by another pump
+/// 2) publish outgoing messages from an egress store to the local/remote broker
+pub struct Pump {
+    client: MqttClient<MessageHandler<WakingMemoryStore>>,
+    client_shutdown: ClientShutdownHandle,
+    publish_handle: PublishHandle,
+    subscriptions: Vec<String>,
+    loader: Rc<RefCell<MessageLoader<WakingMemoryStore>>>,
+    persist: PublicationStore<WakingMemoryStore>,
+    pump_context: PumpContext,
+}
+
+impl Pump {
+    fn new(
+        client: MqttClient<MessageHandler<WakingMemoryStore>>,
+        subscriptions: Vec<String>,
+        loader: Rc<RefCell<MessageLoader<WakingMemoryStore>>>,
+        persist: PublicationStore<WakingMemoryStore>,
+        pump_context: PumpContext,
+    ) -> Result<Self, BridgeError> {
+        let publish_handle = client
+            .publish_handle()
+            .map_err(BridgeError::PublishHandle)?;
+        let client_shutdown = client.shutdown_handle()?;
+
+        Ok(Self {
+            client,
+            client_shutdown,
+            publish_handle,
+            subscriptions,
+            loader,
+            persist,
+            pump_context,
+        })
+    }
+
+    pub async fn subscribe(&mut self) -> Result<(), BridgeError> {
+        self.client
+            .subscribe(&self.subscriptions)
+            .await
+            .map_err(BridgeError::Subscribe)?;
+
+        Ok(())
+    }
+
+    pub async fn run(&mut self, shutdown: Receiver<()>) {
+        debug!(
+            "starting pumps for {} bridge...",
+            self.pump_context.bridge_name
+        );
+
+        let (loader_shutdown, loader_shutdown_rx) = oneshot::channel::<()>();
+        let publish_handle = self.publish_handle.clone();
+        let persist = self.persist.clone();
+        let loader = self.loader.clone();
+        let mut client_shutdown = self.client_shutdown.clone();
+        let bridge_name = self.pump_context.bridge_name.clone();
+        let pump_type = self.pump_context.pump_type.clone();
+
+        // egress pump
+        let f1 = async move {
+            let mut loader_borrow = loader.borrow_mut();
+            let mut receive_fut = loader_shutdown_rx.into_stream();
+
+            debug!(
+                "{} bridge starting egress publication processing for {:?} pump...",
+                bridge_name, pump_type
+            );
+
+            loop {
+                let mut publish_handle = publish_handle.clone();
+                match select(receive_fut.next(), loader_borrow.try_next()).await {
+                    Either::Left((shutdown, _)) => {
+                        debug!("egress pump received shutdown signal");
+                        if let None = shutdown {
+                            error!(message = "unexpected behavior from shutdown signal while signalling bridge pump shutdown")
```

Comment:
Great question. The return type of receive_fut.next() is ```Option<Stream::Item<{unknown}>>```. I was assuming it would return None if something happened to the underlying receiver, but I now realize that is a big assumption.